carapace-server/src/main/java/org/carapaceproxy/server/RequestHandler.java

```diff
@@ -669,11 +669,11 @@ private void addCustomResponseHeaders(HttpObject msg, List<CustomHeader> customH
     }
 
     private boolean releaseConnectionToEndpoint(boolean forceClose, EndpointConnection current) {
-        if (connectionToEndpoint.compareAndSet(current, null)) {
+        if (connectionToEndpoint.get().equals(current)) {
```

Comment:
The first one ho performs releaseConnectionToEndpoint set the connection to null even if the release won't be performed.
The second one, who performs delayed release, cannot release 'cause current != null and the connection won't be never closed